Custom clientconfig

mstergianis opened the pull request:
**What this PR does / why we need it**: 
- Separates functionality of client interface and context switching
- Allows one to use client interfaces not backed by kubeconfigs

**Which issue(s) this PR fixes**
- Fixes no issues

**Special notes for your reviewer**:

**Release note**:
```
release-note
```

cc @dimitrybesson @zoetian

Commit:
Refactor dash.NewRunner to use functional options pattern

Configure cluster package with functional options. This allows for slightly
cleaner invocations of cluster.WithKubeConfig when zero values are to be passed,
and opens the way for more dramatic functional options to come.

Use functional options in CLI.

Add WithClusterClient option for Runner.

Extract KubeContextDecorator interface.
In the cases where kubeconfig contexts make sense, we want a thin wrapper
around cluster.ClientInterface that can handle context switches and report
the set of contexts for the sake of the ContextGenerator. Concrete
implementation is called KubeConfigContextManager.

Remove unused go generate for kubeconfig file loader.

Re-run generation of mock files.

Signed-off-by: Jamie Klassen <jklassen@vmware.com>
Co-authored-by: Michael Stergianis <mstergianis@vmware.com>
Co-authored-by: Zoe Tian <twanru@vmware.com>
Co-authored-by: James Thomson <jamesth@vmware.com>

mstergianis commented:
The "Verify generated code" check appears to be failing because of [removing the Loader interface](https://github.com/vmware-tanzu/octant/pull/1721/files?file-filters%5B%5D=.go&hide-deleted-files=true#diff-2416f88d535bd940665a6d183c33508a77f267716ac70dd9526ca7bbf2c38b9dL17) as the changes we made caused the interface to become unecessary.

GuessWhoSamFoo commented:
Has the mock file been regenerated with `go run build.go go-install && go run build.go generate`?

Commit:
Fix removeFakes in build.go

- remove mock_loader.go generated file

Signed-off-by: Michael Stergianis <mstergianis@vmware.com>

mstergianis commented:
Good catch @GuessWhoSamFoo. I didn't pay enough attention to the diffs and didn't notice that the file was not deleted. I fixed the underlying issue: `removeFakes` in `build.go` was trying to remove `{pkg,internal}/<some package>/fake/fake`

wwitzel3 reviewed (changes requested):
Please clean up the commit history here. We want to avoid things like `WIP` in the mainline history if we can.

wwitzel3 commented:
@GuessWhoSamFoo I created #1744 to add some much needed context around this change.

GuessWhoSamFoo commented on pkg/dash/watcher.go:
```diff
@@ -132,7 +132,7 @@ func (cw *ConfigWatcher) Watch(ctx context.Context) {
 			logger.Infof("shutting down config watcher")
 		case _ = <-cw.FileWatcher.Events():
-			if err := cw.watcherConfig.UseContext(ctx, cw.watcherConfig.ContextName()); err != nil {
+			if err := cw.watcherConfig.UseContext(ctx, cw.watcherConfig.CurrentContext()); err != nil {
```
A potentially interesting edge case (doesn't have to be addressed here) is what should happen if a config is loaded but becomes invalid (e.g. deleted) while Octant is running.

GuessWhoSamFoo commented on internal/event/context.go:
```diff
@@ -27,18 +24,21 @@ type kubeContextsResponse struct {
 	kcg := &ContextsGenerator{
-		ConfigLoader: kubeconfig.NewFSLoader(),
-		DashConfig:   dashConfig,
+		KubeContextStore: kubeContextStore,
```
Make sure the context is updating as the kubeconfig changes from the watcher. Current behavior has the context updating in Octant via `kubectl config use-context`. A similar case also holds if additional contexts are added though tools like kind.

Additionally, an instance of Octant running via `FromClientConfig` should not be looking at the filesystem.

GuessWhoSamFoo commented:
I haven't looked into why - these changes also resurface an older issue when switching contexts
```
2020-12-17T13:43:45.562-0800    ERROR   api/content_manager.go:154      generate content     {"client-id": "ccdafa4b-40b0-11eb-a4ed-
8c16455fb1b6", "err": "generate content: preferred resources: unable to retrieve the complete list of server APIs: 
metrics.k8s.io/v1beta1: the server is currently unable to handle the request", "content-path": "overview/namespace/kube-system"
```

Commit:
Restore functionality to infer current context on fs change

Signed-off-by: Michael Stergianis <mstergianis@vmware.com>
Signed-off-by: Vikram Yadav <yvikram@vmware.com>

mstergianis reviewed (commented):
> I haven't looked into why - these changes also resurface an older issue when switching contexts
> 
> ```
> 2020-12-17T13:43:45.562-0800    ERROR   api/content_manager.go:154      generate content     {"client-id": "ccdafa4b-40b0-11eb-a4ed-
> 8c16455fb1b6", "err": "generate content: preferred resources: unable to retrieve the complete list of server APIs: 
> metrics.k8s.io/v1beta1: the server is currently unable to handle the request", "content-path": "overview/namespace/kube-system"
> ```
> 
> EDIT: I actually do; we need to still generate content even if not in the list of preferred server resource.

I'm looking into this now, but am unable to reproduce by switching contexts in the UI or via `kubectl`. Are there any other steps needed to reproduce?

mklanjsek commented:
> Are there any other steps needed to reproduce?

@mstergianis To repro that issue, you need a failing resource on your cluster  - in this case `metrics.k8s.io/v1beta1` had missing endpoints.  That could be hard to generate, in that case maybe @wwitzel3 can give you access to our dev cluster.

mstergianis commented:
Thanks @mklanjsek, will do

wwitzel3 commented:
The metrics issue was not re-introduced in this PR, it was re-introduced here: https://github.com/vmware-tanzu/octant/commit/dcb101861ff11cd521317464d29d81ed1b9047eb

wwitzel3 commented:
Looks like the current failing test on here is a legitimate one? @mstergianis

mstergianis commented:
Okay, so I've moved off of this track of work, but have brought @xtreme-vikram-yadav and @weresch up to speed on why the tests are failing.

Before the tests can be fixed though I want to confirm what the intended behaviour is for octant to monitor filesystem changes to the kubeconfig.

Before PR:
1. octant starts up
2. octant begins monitoring user's kubeconfig in https://github.com/vmware-tanzu/octant/blob/master/pkg/dash/watcher.go#L126-L142
3. when the user switches context in octant via the websocket, a non empty string is written to this in memory field https://github.com/vmware-tanzu/octant/blob/915279cbe3431884a00c50a85b87b597e750a425/internal/config/dash.go#L122
4. octant no longer switches context when the user changes their kubeconfig on fs because the watcher passes in the current octant context in to the `UseContext` method

After PR:
1. octant starts up
2. octant begins monitoring user's kubeconfig in https://github.com/mstergianis/octant/blob/custom-clientconfig/pkg/dash/watcher.go#L126-L142
3. when the user switches contexts via websocket, the watcher continues to pass the empty string into `UseContext` so the kubeconfig is always parsed, and the octant context is always synced with the filesystem

In light of [this](https://github.com/vmware-tanzu/octant/issues/1793) conversation with @GuessWhoSamFoo, we are unsure if the behaviour to ignore changes to the filesystem are intentional or not. This impacts if we fix the tests or the code.

wwitzel3 commented:
We do not want Octant to persist context switching changes to the filesystem.

mstergianis commented:
Spoke to Wayne offline, we're going to modify the `KubeContextDecorator` so it behaves as closely as possible to the way Live switched contexts in the past.

wwitzel3 commented:
Do you have any more questions about the behavior or the failing test?

wwitzel3 commented:
How's it going on this?

weresch commented:
Hi Wayne, I've picked up this PR in the meantime. I've traced the behaviour of the original code to see how kube context is passed around. Now tracing the behaviour of the PR code so it can be updated to replicate the behaviour of the original code.

weresch committed:
Restoring the behaviour where Octant context is updated to file system's
kubeconfig if it is changed unless the user has chosen a context in Octant UI.

Added SetContextChosenInUI to dash config to explicitly track when the user
has chosen their Octant context in the UI.

Added UseFSContext constant to be used to indicate a context switch to the
file system Kubeconfig context.

Signed-off-by: Vikram Yadav <vyadav@vmware.com>
